doc/source/tutorial.rst

```diff
@@ -0,0 +1,284 @@
+
+  x_id = ray.put([1, 2, 3])
+  ray.get(x_id)  # [1, 2, 3]
+
+If the remote object corresponding to the object ID ``x_id`` has not been created
+yet, the command ``ray.get(x_id)`` will wait until the remote object has been
+created.
+
+A very common use case of ``ray.get`` is to get a list of object IDs. In this
+case, you can call ``ray.get(object_ids)`` where ``object_ids`` is a list of object
+IDs.
+
+.. code-block:: python
+
+  result_ids = [ray.put(i) for i in range(10)]
+  ray.get(result_ids)  # [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+Asynchronous Computation in Ray
+-------------------------------
+
+Ray enables arbitrary Python functions to be executed asynchronously. This is
+done by designating a Python function as a **remote function**.
+
+For example, a normal Python function looks like this.
+
+.. code-block:: python
+
+  def add1(a, b):
+    return a + b
+
+A remote function looks like this.
+
+.. code-block:: python
+
+  @ray.remote
+  def add2(a, b):
+    return a + b
+
+Remote functions
+~~~~~~~~~~~~~~~~
+
+Whereas calling ``add1(1, 2)`` returns ``3`` and causes the Python interpreter to
+block until the computation has finished, calling ``add2.remote(1, 2)``
+immediately returns an object ID and creates a **task**. The task will be
+scheduled by the system and executed asynchronously (potentially on a different
+machine). When the task finishes executing, its return value will be stored in
+the object store.
+
+.. code-block:: python
+
+  x_id = add2.remote(1, 2)
+  ray.get(x_id)  # 3
+
+The following simple example demonstrates how asynchronous tasks can be used
+to parallelize computation.
+
+.. code-block:: python
+
+  import time
+
+  def f1():
+    time.sleep(1)
+
+  @ray.remote
+  def f2():
+    time.sleep(1)
+
+  # The following takes ten seconds.
+  [f1() for _ in range(10)]
+
+.. code-block:: python
+
+  # The following takes one second (assuming the system has at least ten CPUs).
+  ray.get([f2.remote() for _ in range(10)])
+
+There is a sharp distinction between *submitting a task* and *executing the
+task*. When a remote function is called, the task of executing that function is
+submitted to a local scheduler, and object IDs for the outputs of the task are
+immediately returned. However, the task will not be executed until the system
+actually schedules the task on a worker. Task execution is **not** done lazily.
+
+**When a task is submitted, each argument may be passed in by value or by object
+ID.** For example, these lines have the same behavior.
+
+.. code-block:: python
+
+  add2.remote(1, 2)
+  add2.remote(1, ray.put(2))
+  add2.remote(ray.put(1), ray.put(2))
+
+Remote functions never return actual values, they always return object IDs.
+
+When the remote function is actually executed, it operates on Python objects.
+That is, if the remote function was called with any object IDs, the Python
+objects corresponding to those object IDs will be retrieved and passed into the
+actual execution of the remote function.
```

Comment:
I think we can remove most of this text and add a line after the sentence above ("When a task is submitted...") saying that the system retrieves the argument's value from the object store if it is an object ID.